# CSE 220 Paper Review <br/> Darwin - A Genomics co-processor for long read assembly


## Summary

This paper describes a co-processor that optimizes DNA assembly for long reads (average string length
of 10000 characters) where previous work optimizes for short reads (average string length of 100
characters). Background of the workload is given: assembly is typically a dynamic programming algorithm,
but an optimization finds exact matches of a short length, and extends string matches on either side of
the matched substring (seed) of a read (query string) compared to the reference assembly (reference
string). In addition to handling reference-based assembly (match DNA fragments to a pre-assembled genome),
the co-processor is also able to handle other, similar workloads--homology search, de-novo assembly (assembly
of a genome without a reference), and others. Then, the hardware design is described: the algorithm for the
filtering step (matching short substrings), D-SOFT, is addressed using a network-on-chip (NoC) with a butterfly
design; and the alignment step, GACT, is addressed using a systolic array of processing elements (PE), where
each PE has a 2kb SRAM for traceback (optimal solution in dynamic programming matrix. The extend step uses
wavefront parallelism to fill the dynamic programming matrix efficiently. The filtering step has copies of
query strings, reference strings, and all parameters in each DRAM channel in order to maximize the utilization
of each DRAM.

There are a few different aspects of evaluation included in the paper. One that I find interesting is the
area and power breakdown of Darwin components, which also mentions that the critical path of Darwin is 1.18 ns
in the table caption. There is also a table of error rates for 3 different sequencing technologies (pacbio,
nanopore 2D, nanopore 1D) and how well they were aligned for two parameters: T (tiles, or entries in the
dynamic programming table) and O. For 1000 length reads up to 10000, GACT (Darwin) performs faster than GACT
(software).


## What are the paper's strengths?



## What are the paper's weaknesses?



## Suggestions to improve paper



## Impact and more recent work



## Citation information

    @inproceedings{conf:darwin-coprocessor,
        author    = {Yatish Turakhia and Gill Bejerano and William J. Dally},
        title     = {Darwin - A Genomics Co-processor Provides up to 15000x acceleration on long
                     read assembly},
        series    = {ASPLOS '18},
        year      = {2018},
        location  = {Williamsburg, VA, USA},
        pages     = {199--213},
        numpages  = {15},
        url       = {http://doi.acm.org/10.1145/3173162.3173193},
        doi       = {10.1145/3173162.3173193},
        acmid     = {3173193},
        publisher = {ACM},
    }
