# Programmable Access Libraries for Bioinformatics (Work in Progress)

We are interested in attempting to improve performance of data access for life science use cases,
specifically [the Human Cell Atlas (HCA)][hca].

## Ideas and Tasks
Tasks:

* collaborative filtering between OSDs given:
    * an ontology (for knowing the collaborative hierarchy)
    * partial aggregates (sort a sub-region then coordinate along intervals)
    * incremental computation
    * view maintenance
* incremental/hierarchical DBSCAN (or whatever clusting algorithm)
* try transform of gene expression data (column-wise rather than row-wise)

Ideas:
* The use of an ontology (see [Ontologial Hierarchical Clustering][ohclust]):

      Using an ontology can enable data locality and incremental clustering.

* The exchange of partial aggregates between OSDs:

      Collaborative OSD partial aggregates allows OSDs to communicate intermediate results
      for improving workloads such as sorts and percentiles and essentially enable
      divide-and-conquer algorithms.
   
* The effects of data layout of gene expression data (the data to be analyzed in parallel):

      By changing the layout of gene expression data to group all expressions of a gene,
      it may be possible to speedup comparisons for the same gene. Likely, it would be ideal
      to write a normalization function in the storage system so that writes of a new sample
      could normalize gene expressions so that analysis over gene expressions can be done
      without having to access values for other genes.


<!-- Resources -->
[hca]:     https://www.humancellatlas.org/
[ohclust]: https://ieeexplore.ieee.org/document/6753971
